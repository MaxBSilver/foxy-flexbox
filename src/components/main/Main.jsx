import React from "react";
import Answer from "../answer/Answer";
import Board from "../board/Board";
import Prompt from "../prompt/Prompt";


class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      question: "",
      roundNumber: 0,
      userGuessOne: "",
      userGuessTwo: "",
      difficulty: "",
      display: "",
      flexDirection: "",
      isLoading: true
    };
  }

  componentDidMount = () => {
    this.determineQuestion();
  };

  determineQuestion = () => {
    this.selectQuestion();
  };

  selectQuestion = () => {
    const { roundNumber } = this.state;
    this.setState({
      question: this.props.data[roundNumber],
      userGuessOne: "",
      userGuessTwo: "",
      userGuessThree: "",
      isLoading: false
    });
  };

  incrementRound = () => {
    let incrementedRound = this.state.roundNumber;
    incrementedRound = incrementedRound += 1;
    if (incrementedRound < 20) {
      this.setState({ roundNumber: incrementedRound }, () => {
        this.determineQuestion();
      });
    }
  };

  decrementRound = () => {
    let decrementedRound = this.state.roundNumber;
    decrementedRound = decrementedRound -= 1;
    if (decrementedRound >= 0) {
      this.setState({ roundNumber: decrementedRound }, () => {
        this.determineQuestion();
      });
    }
  };

  updateUserAnswer = guess => {
    const { acVal, jcVal } = guess;
    this.setState({
      userGuessOne: jcVal,
      userGuessTwo: acVal
    });
  };

  render() {
    const { question, userGuessOne, userGuessTwo, roundNumber } = this.state;
    return !this.state.isLoading ? (
      <main>
        <section className="left--container">
          <Prompt
            {...question}
            determineQuestion={this.determineQuestion}
            incrementRound={this.incrementRound}
            decrementRound={this.decrementRound}
          />
          <Answer
            {...question}
            updateUserAnswer={this.updateUserAnswer}
            determineQuestion={this.determineQuestion}
            incrementRound={this.incrementRound}
          />
        </section>
        <section className="right--container">
          <Board
            justifyContent={question.justifyContent}
            userGuessOne={userGuessOne}
            alignContent={question.alignContent}
            userGuessTwo={userGuessTwo}
            prompts={question.prompt}
            roundNumber={roundNumber}
            difficulty={question.difficulty}
            display={question.display}
            flexDirection={question.flexDirection}
          />
        </section>
      </main>
    ) : (
      <div />
    );
  }
}

export default Main;
