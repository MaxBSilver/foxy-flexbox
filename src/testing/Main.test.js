import React from 'react';
import Main from '../components/main/Main';
import { shallow } from 'enzyme';
describe("Main", () => {
  let mockData = {
    data: [
      {
        id: 1,
        difficulty: "easy",
        display: "flex",
        flexDirection: "row",
        prompt: "justify-content",
        justifyContent: "center",
        alignContent: "flex-end"
      },
      {
        id: 2,
        difficulty: "easy",
        display: "flex",
        flexDirection: "row",
        prompt: "justify-content",
        justifyContent: "flex-end",
        alignContent: "flex-end"
      },
      {
        id: 3,
        difficulty: "easy",
        display: "flex",
        flexDirection: "row",
        prompt: "justify-content",
        justifyContent: "flex-end",
        alignContent: "center"
      },
      {
        id: 4,
        difficulty: "easy",
        display: "flex",
        flexDirection: "row",
        prompt: "justify-content",
        justifyContent: "flex-start",
        alignContent: "flex-end"
      }
    ]
  };
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Main {...mockData} />);
  });
  it("should match the snapshot with all data passed in", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should increment round", () => {
    expect(wrapper.state("roundNumber")).toEqual(0);
    wrapper.instance().incrementRound();
    expect(wrapper.state("roundNumber")).toEqual(1);
    wrapper.instance().incrementRound();
    expect(wrapper.state("roundNumber")).toEqual(2);
  });
  it("should decrement round", () => {
    expect(wrapper.state("roundNumber")).toEqual(0);
    wrapper.instance().incrementRound();
    expect(wrapper.state("roundNumber")).toEqual(1);
    wrapper.instance().decrementRound();
    expect(wrapper.state("roundNumber")).toEqual(0);
  });
  it("should not decrement round under 0", () => {
    expect(wrapper.state("roundNumber")).toEqual(0);
    wrapper.instance().decrementRound();
    expect(wrapper.state("roundNumber")).toEqual(0);
  });
  it("should update the question when the round is incremented or decremented", () => {
    expect(wrapper.state("question")).toEqual({
      alignContent: "flex-end",
      difficulty: "easy",
      display: "flex",
      flexDirection: "row",
      id: 1,
      justifyContent: "center",
      prompt: "justify-content"
    });
    wrapper.instance().incrementRound();
    expect(wrapper.state("question")).toEqual({
      id: 2,
      difficulty: "easy",
      display: "flex",
      flexDirection: "row",
      prompt: "justify-content",
      justifyContent: "flex-end",
      alignContent: "flex-end"
    });
    wrapper.instance().decrementRound();
    expect(wrapper.state("question")).toEqual({
        alignContent: "flex-end",
        difficulty: "easy",
        display: "flex",
        flexDirection: "row",
        id: 1,
        justifyContent: "center",
        prompt: "justify-content"
      });
  });
  it("should update user answer", () => {
    expect(wrapper.state("userGuessOne")).toEqual('');
    expect(wrapper.state("userGuessTwo")).toEqual('');
    wrapper.instance().updateUserAnswer({jcVal : 'center', acVal : 'flex-end'});
    expect(wrapper.state("userGuessOne")).toEqual('center');
    expect(wrapper.state("userGuessTwo")).toEqual('flex-end');
  })
  it("should update question on load", () => {
    expect(wrapper.state("question")).toEqual({"alignContent": "flex-end", "difficulty": "easy", "display": "flex", "flexDirection": "row", "id": 1, "justifyContent": "center", "prompt": "justify-content"});
  })
});