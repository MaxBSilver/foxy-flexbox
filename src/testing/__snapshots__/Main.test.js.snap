// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should match the snapshot with all data passed in 1`] = `
<main>
  <section
    className="left--container"
  >
    <Prompt
      alignContent="flex-end"
      decrementRound={[Function]}
      determineQuestion={[Function]}
      difficulty="easy"
      display="flex"
      flexDirection="row"
      id={1}
      incrementRound={[Function]}
      justifyContent="center"
      prompt="justify-content"
    />
    <Answer
      alignContent="flex-end"
      determineQuestion={[Function]}
      difficulty="easy"
      display="flex"
      flexDirection="row"
      id={1}
      incrementRound={[Function]}
      justifyContent="center"
      prompt="justify-content"
      updateUserAnswer={[Function]}
    />
  </section>
  <section
    className="right--container"
  >
    <Board
      alignContent="flex-end"
      difficulty="easy"
      display="flex"
      flexDirection="row"
      justifyContent="center"
      prompts="justify-content"
      roundNumber={0}
      userGuessOne=""
      userGuessThree=""
      userGuessTwo=""
    />
  </section>
</main>
`;
